src/shared/utils/url.ts

```diff
@@ -10,21 +9,40 @@ export const getLocationForPathname = (pathname: string) => {
   return found?.[0];
 };
 
-/**
-Wrapper around query-string's parse function which takes the output of this and
-returns only the first value of a parameter if it is an array.
+type QueryStringParamsRecord = Record<
+  string,
+  string | string[] | number | number[] | boolean | undefined | null
+>;
+export type QueryStringArg = string | QueryStringParamsRecord | URLSearchParams;
 
-@param query - The query string to parse.
-@param options - Options to pass query-string's parse 
-*/
-export const parseQueryString = (
-  query: string,
-  options?: queryString.ParseOptions
-) => {
-  const parsed = queryString.parse(query, options);
-  const parsedWithoutArrayValues: { [key: string]: string | null } = {};
-  for (const [key, value] of Object.entries(parsed)) {
-    parsedWithoutArrayValues[key] = Array.isArray(value) ? value[0] : value;
+export const stringifyQuery = (...args: QueryStringArg[]) => {
+  // This returns a query string by iterating over the args and creating a combined
+  // URLSearchParams instance. If a parameter key already exists it will be overwritten
+  // by the new value. If the new value is undefined or null then the parameter will
+  // be removed. Arrays will be returned as comma separated strings.
+  const combined = new URLSearchParams();
+  for (const arg of args) {
+    const iter =
+      (typeof arg === 'string' && new URLSearchParams(arg)) ||
+      (arg instanceof URLSearchParams && arg) ||
+      (typeof arg !== 'undefined' && Object.entries(arg)) ||
+      [];
+    for (const [k, v] of iter) {
+      if (typeof v !== 'undefined' && v !== null) {
+        combined.set(k, v.toString());
+      } else if (combined.has(k)) {
+        combined.delete(k);
+      }
+    }
   }
-  return parsedWithoutArrayValues;
+  combined.sort();
+  return combined.toString();
+};
+
+export const stringifyUrl = (base: string, ...args: QueryStringArg[]) =>
+  `${base}?${stringifyQuery(...args)}`;
```

Comment:
I would maybe stringify the query first, then if it's truthy I would happen both the question mark and the stringified query (instead of having the URL finish with a strange `?` only